src/Workspaces/Remote/Core/Serialization/ImmutableSegmentedListFormatter.cs

```diff
@@ -0,0 +1,60 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using MessagePack;
+using MessagePack.Formatters;
+using Microsoft.CodeAnalysis.Collections;
+
+namespace Microsoft.CodeAnalysis.Remote;
+
+internal sealed class ImmutableSegmentedListFormatter<T> : IMessagePackFormatter<ImmutableSegmentedList<T>>
+{
+    ImmutableSegmentedList<T> IMessagePackFormatter<ImmutableSegmentedList<T>>.Deserialize(ref MessagePackReader reader, MessagePackSerializerOptions options)
+    {
+        if (reader.TryReadNil())
+            return default;
+
+        var len = reader.ReadArrayHeader();
+        if (len == 0)
+            return ImmutableSegmentedList<T>.Empty;
+
+        var formatter = options.Resolver.GetFormatterWithVerify<T>();
+
+        // TODO: Should there be a CreateBuilder that takes in an initial capacity
+        // similar to what ImmutableArray allows?
```

Comment:
Yes, there should. In the meantime, you can use `SegmentedCollectionsMarshal.AsImmutableSegmentedList` if you want.

Reply:
I have a branch ready to submit for this as soon as #62193 merges.